src/groups/mqb/mqbblp/mqbblp_relayqueueengine.cpp

```diff
@@ -1682,17 +1676,472 @@ bool RelayQueueEngine::subscriptionId2upstreamSubQueueId(
+             subscriptions.begin();
+         cit != subscriptions.end();
+         ++cit) {
+        const bmqp::SubQueueInfo& subscription = *cit;
+
+        unsigned int subQueueId;
+
+        if (!subscriptionId2upstreamSubQueueId(msgGUID,
+                                               &subQueueId,
+                                               subscription.id())) {
+            continue;  // CONTINUE
+        }
+
+        PushStream::Apps::iterator itApp = d_pushStream.d_apps.find(
+            subQueueId);
+        if (itApp == d_pushStream.d_apps.end()) {
+            AppsMap::const_iterator cit = d_apps.find(subQueueId);
+
+            if (cit == d_apps.end()) {
+                BMQ_LOGTHROTTLE_ERROR()
+                    << "#QUEUE_UNKNOWN_SUBSCRIPTION_ID "
+                    << "Remote queue: " << d_queueState_p->uri()
+                    << " (id: " << d_queueState_p->id()
+                    << ") discarding a PUSH message for guid " << msgGUID
+                    << ", with unknown App Id " << subscription.id();
+                continue;  // CONTINUE
+            }
+
+            itApp = d_pushStream.d_apps.emplace(subQueueId, cit->second).first;
+        }
+        else if (!checkForDuplicate(itApp->second.d_app.get(), msgGUID)) {
+            continue;  // CONTINUE
+        }
+
+        PushStream::Element* element = d_pushStream.create(subscription,
+                                                           itGuid,
+                                                           itApp);
+
+        d_pushStream.add(element);
+        ++count;
+    }
+
+    if (count) {
+        // Pass correct ref count
+        attributes->setRefCount(count);
+        storePush(attributes, msgGUID, appData, false);
+    }
+    return count;
+}
+
+bool RelayQueueEngine::checkForDuplicate(const App_State*         app,
+                                         const bmqt::MessageGUID& msgGUID)
+{
+    // Check the storage if this is duplicate PUSH
+    // Currently, only Proxies can do this because they clear their storage
+    // once all readers are gone.  Therefore, a Proxy can detect duplicate PUSH
+    // when there is a client (potentially) processing the same PUSH.
+    // Replicas do not clear their (replicated) storage.  They do not check if
+    // there is a client (potentially) processing the same PUSH.
+    // Also, Proxies use InMemoryStorage which relies on PUSH uniqueness for
+    // calculating the refCount.
+    // (Replicas receive the refCount by replication).
+
+    if (d_queueState_p->domain()->cluster()->isRemote()) {
+        d_realStorageIter_mp->reset(msgGUID);
+        if (!d_realStorageIter_mp->atEnd()) {
+            mqbi::AppMessage& appState = d_realStorageIter_mp->appMessageState(
+                app->ordinal());
+
+            if (!appState.isPushing()) {
+                appState.onPush();
+            }
+            else {
+                BMQ_LOGTHROTTLE_INFO()
```

Comment:
These might be good metrics to have down the road.  We can scrape them from logs, but it would be interesting to see, in production, how frequent duplicate PUSHs are.